cmd/attach/policy/cmd.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright (c) 2024 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package policy
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/rosa/pkg/aws/tags"
+	"github.com/openshift/rosa/pkg/interactive"
+	"github.com/openshift/rosa/pkg/ocm"
+	"github.com/openshift/rosa/pkg/rosa"
+)
+
+const (
+	use     = "policy"
+	short   = "Attach policy to a role"
+	long    = "Attach existing policy to role in the user’s AWS account"
+	example = `  # Attach policy <policy_arn> to role <role_name>
+  rosa attach policy --policy-arn=<policy_arn> --role-arn=<role_name>`
+)
+
+var args struct {
+	policyArn string
+	roleName  string
+}
+
+func NewAttachPolicyCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     use,
+		Short:   short,
+		Long:    long,
+		Example: example,
+		Args:    cobra.NoArgs,
+		Run:     rosa.DefaultRunner(rosa.RuntimeWithOCMAndAWS(), AttachPolicyRunner()),
+	}
+
+	flags := cmd.Flags()
+
+	flags.StringVarP(
+		&args.policyArn,
+		"policy-arn",
+		"p",
+		"",
+		"Policy arn of the policy to be attached to the specified role (required).",
+	)
+	flags.StringVarP(
+		&args.roleName,
+		"role-name",
+		"r",
+		"",
+		"Role name of the role to attach the specified policy (required).",
+	)
+	cmd.MarkFlagRequired("policy-arn")
+	cmd.MarkFlagRequired("role-name")
+
+	interactive.AddModeFlag(cmd)
+	return cmd
+}
+
+func AttachPolicyRunner() rosa.CommandRunner {
+	return func(_ context.Context, r *rosa.Runtime, cmd *cobra.Command, _ []string) error {
+		mode, err := interactive.GetMode()
+		if err != nil {
+			return err
+		}
+
+		defer r.Cleanup()
+
+		policyArn := args.policyArn
+		_, err = r.AWSClient.IsPolicyExists(policyArn)
```

Comment:
@marcolan018 Can I make a suggestion of defining an `ArbitraryPolicyService` and having a few simple methods that hide all of this implementation and flow detail from the `Runner` implementation. 

So for example, how about a service interface defined like so:

```
type ArbitraryPolicyService interface {
   AutoAttachArbitraryPolicy(roleName, policyArn) error
   ManualAttachArbitraryPolicy(roleName, policyArn) (string, error)
}
```

You could then create an implementation of it that requires the `AWSClient` and `OCMClient` which you get an instance of via a creator function:

```
arbitraryPolicyService := policies.NewArbitraryPolicyService(r.AWSClient, r.OCMClient)
```

That would then shorten the logic in this `Runner` to _something_ like:

```
mode, err := interactive.GetMode()
if err != nil {
	return err
}

// Determine if interactive mode is needed
if !interactive.Enabled() && !cmd.Flags().Changed("mode") {
	interactive.Enable()
}

if interactive.Enabled() {
	mode, err = interactive.GetOptionMode(cmd, mode, "Attach policy mode")
	if err != nil {
		return fmt.Errorf("Expected a valid attach policy mode: %s", err)
	}
}

arbitraryPolicyService := policies.NewArbitraryPolicyService(r.AWSClient, r.OCMClient)

switch mode {
	case interactive.ModeAuto:
		return arbitraryPolicyService.AutoAttachArbitraryPolicy(roleName, policyArn)
	case interactive.ModeManual:
		cmd, err := arbitraryPolicyService.ManualAttachArbitraryPolicy(roleName, policyArn)
                if err != nil {
                   return err
                }
               r.Reporter.Infof("Run the following command to attach the policy: %s", cmd)
	default:
		return fmt.Errorf("Invalid mode. Allowed values are %s", interactive.Modes)
	}
return nil
```

The main benefit I see of this is a cleaner separation of concerns and ease of testing. The `Runner` is purely focused on doing CLI-stuff e.g getting input from the user and printing output. The actual logic of how this works in encapsulated in the service and if we ever need to change/extend it, the runner doesn't need to be altered.

Reply:
make the code refactoring